### PR TITLE
Drop sqlglot for generating CTEs

### PR DESCRIPTION
**Pull request by tonykploomber**
## Describe your changes

In the original issue, we need to do:
1. Need some research to see what dialects support backticks, and make a list
2. Remove the sqlglot & add condition on using non-backticks or backticks jinja template

Result:
1. We can check if the implementation of the corresponding dialect in sqlglot has IDENTIFIERS = ["`"] attribute, which is confirmed by sqlglot team
2. We prepare two template in this PR. One is using `"`, another one is using ` (backtick), we will dynamic choose the right template by checking the implementation of the sqlglot.dialect class

## Issue number

Closes https://github.com/ploomber/jupysql/issues/307

## Checklist before requesting a review

- [ ] Performed a self-review of my code
- [ ] Formatted my code with [`pkgmt format`](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#lintingformatting)
- [ ] Added [tests](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#testing) (when necessary).
- [ ] Added documentation in the [docstring](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#documenting-changes-and-new-features) and [changelog](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--318.org.readthedocs.build/en/318/

### CODE REVIEW

File: src/sql/connection.py [hunk 1]
```diff
@@ -365,14 +369,55 @@ def _get_curr_connection_info(cls):
...
+        cur_dialect = cls._get_curr_sqlglot_dialect()
+        if not cur_dialect:
+            return False
         try:
-            write_dialect = DIALECT_NAME_SQLALCHEMY_TO_SQLGLOT_MAPPING.get(
-                connection_info["dialect"], connection_info["dialect"]
+            return (
+                "`" in sqlglot.Dialect.get_or_raise(cur_dialect).Tokenizer.IDENTIFIERS
             )
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (tonykploomber) [reply src/sql/connection.py]: 1. We might still need to `except Exception as e`, to catch the original exception
2. We use `except Exception` because the `e` might be the three cases: 
  - ValueError, when dialect isn't defined in sqlglot
  - AttributeError, when dialect is defined, but no Tokenizer class declared inside
  - TypeError, when dialect && Tokenizer class are defined, but no IDENTIFIERS list declared inside


I rewrote in [Fix: not re-raise Exception in _is_curr_dialect_support_backtick](https://github.com/ploomber/jupysql/pull/318/commits/5100677b5c0f26349e3ca389ee59038c6640b755), but I think original way I did is wrong since it will re-raise new exception, also I added some test cases for this.

Please LMK if this you want 

Ref: [How do I raise the same Exception with a custom message in Python?](https://stackoverflow.com/questions/9157210/how-do-i-raise-the-same-exception-with-a-custom-message-in-python)

[the remark this answers is not in the document]
Comment (tonykploomber) [reply src/sql/connection.py]: Another approach is we handle those three exception cases in _is_curr_dialect_support_backtick directly, and return False when any of those happens

Implementation: [Maybe just return false?](https://github.com/ploomber/jupysql/pull/318/commits/9bd9aa1eee2fa98d09e54af2330ba95418414ced)

LMK which one you think is better?

[the remark this answers is not in the document]
Comment (tonykploomber) [reply src/sql/connection.py]: Because I think the purpose of `_is_curr_dialect_support_backtick` function is just return if the dialect can support backtick (`) syntax identifier, the caller function just need to know True / False is enough?

Since this function will treat all exception cases as not support, we don't have to handle the exception in the caller function since it's not support anyway

File: src/sql/connection.py [hunk 2]
```diff
@@ -365,14 +369,55 @@ def _get_curr_connection_info(cls):
...
+        dialect by sqlglot
+
+        Args:
+            query (str): Original SQL clause
+
+        Returns:
+            str: SQL clause that's compatible to current connected dialect
+        """
+        write_dialect = cls._get_curr_sqlglot_dialect()
+        try:
```
[the remark this answers is not in the document]
Comment (tonykploomber) [reply src/sql/connection.py:420]: Good question, we discussed [this before](https://github.com/ploomber/jupysql/pull/164#discussion_r1119452604)

The main reason we don't want just raise the exception is this section relies on external library sqlglot, since the function here if to transpile the original query if possible, we want to safely return the original query when external library has any issues

File: src/tests/test_connection.py [hunk 3]
```diff
@@ -45,6 +46,86 @@ def test_get_curr_connection_info(mock_postgres):
...
+        ),
+    ],
+)
+def test_get_curr_sqlglot_dialect(
+    monkeypatch, sqlalchemy_connection_info, expected_sqlglot_dialect
+):
+    monkeypatch.setattr(
+        Connection, "_get_curr_connection_info", lambda: sqlalchemy_connection_info
+    )
+    monkeypatch.setattr(
```
[the remark this answers is not in the document]
Comment (tonykploomber) [reply src/tests/test_connection.py:99]: Do you mean this?

```
def test_get_curr_sqlglot_dialect(
    monkeypatch, sqlalchemy_connection_info, expected_sqlglot_dialect
):
    mock_connection = Mock()
    mock_connection._get_curr_connection_info = lambda: sqlalchemy_connection_info
    mock_connection.DIALECT_NAME_SQLALCHEMY_TO_SQLGLOT_MAPPING = {"sqlalchemy_mock_dialect_name": "sqlglot_mock_dialect"}
    assert mock_connection._get_curr_sqlglot_dialect() == expected_sqlglot_dialect
```

Got some questions:
If we mock the original `sql.connection`, how we test the `_get_curr_sqlglot_dialect` function inside since it's parent class is being mocked?

File: src/tests/test_connection.py [hunk 4]
```diff
@@ -45,6 +46,86 @@ def test_get_curr_connection_info(mock_postgres):
...
+):
+    monkeypatch.setattr(Connection, "_get_curr_sqlglot_dialect", lambda: cur_dialect)
+    assert Connection._is_curr_dialect_support_backtick() == expected_support_backtick
+
+
+def test_is_curr_dialect_support_backtick_sqlglot_missing_dialect(monkeypatch):
+    monkeypatch.setattr(
+        Connection, "_get_curr_sqlglot_dialect", lambda: "something_weird_dialect"
+    )
+    with pytest.raises(ValueError):
```
[the remark this answers is not in the document]
Comment (tonykploomber) [reply src/tests/test_connection.py]: We should keep `ValueErrors`, this is from sqlglot.Dialect.[get_or_raise](https://github.com/tobymao/sqlglot/blob/main/sqlglot/dialects/dialect.py#L136), we can't change it since this happens when there is no dialect implemented

File: src/tests/test_connection.py [hunk 5]
```diff
@@ -2,9 +2,12 @@
 from unittest.mock import Mock, patch
 import pytest
 from sqlalchemy.engine import Engine
+import sql.connection
 from sql.connection import Connection
 from IPython.core.error import UsageError
 
+# import sqlglot
```
Comment (yafimvo) [inline src/tests/test_connection.py:9]: We can remove this line

File: src/sql/store.py [hunk 6]
```diff
@@ -98,14 +99,23 @@ def __init__(self, store: SQLStore, query: str, with_: Iterable = None):
...
+        with_template_backtick = Template(
+            """WITH{% for name in with_ %} `{{name}}` AS ({{saved[name]._query}})\
+{{ "," if not loop.last }}{% endfor %}{{query}}"""
+        )
+        is_use_backtick_template = (
+            sql.connection.Connection._is_curr_dialect_support_backtick()
+        )
+        with_all = _get_dependencies(self._store, self._with_)
+        print("with_all", with_all)
+        template = with_template_backtick if is_use_backtick_template else with_template
```
Comment (yafimvo) [inline src/sql/store.py:115]: As I understand, this PR should handle 2 templates and dynamically choose the right one. I guess there are more possible templates we might handle in the future, so to make it easier to add more templates in the future I think we can export the template logic to another function i.e : `_get_template()` which returns the right template for a given connection.
```
def __str__(self) -> str:
        template = _get_template()
        return template.render(
            query=self._query, saved=self._store._data, with_=with_all
        )
```

as for the implementation of choosing the right template, I think at this point, when we have only 2 possible templates - it's easy to use the if statement and manage 2 queries, so we can leave it as is.

But we need to keep in mind that if we'll want to add another template we'll have to implement another function like `is_use_backtick_template` i.e : `is_use_something_new_template`.

So, I think if there is not much effort in making it more scalable to handle future templates - I would do it (maybe a dynamic query builder with given identifiers).
If not - I think we can add a comment explaining what are the steps to create a new template.

**Review by yafimvo — CHANGES_REQUESTED** [review]
left some comments.

File: src/sql/store.py (hunk 6 above)
Comment (tonykploomber) [reply src/sql/store.py]: Yes, I agree your point!
We can have the helper class to handle this in the future if there are more templates needed.

File: src/sql/store.py [hunk 7]
```diff
@@ -98,14 +99,22 @@ def __init__(self, store: SQLStore, query: str, with_: Iterable = None):
...
-        return ""
+        with_clause_template = Template(
+            """WITH{% for name in with_ %} "{{name}}" AS ({{saved[name]._query}})\
+{{ "," if not loop.last }}{% endfor %}{{query}}"""
+        )
+        with_clause_template_backtick = Template(
+            """WITH{% for name in with_ %} `{{name}}` AS ({{saved[name]._query}})\
+{{ "," if not loop.last }}{% endfor %}{{query}}"""
+        )
+        is_use_backtick = sql.connection.Connection.is_use_backtick_template()
```
Comment (yafimvo) [inline src/sql/store.py:110]: For readability and to facilitate future changes, I think we can add some comments here. Also, I see there is no difference between the 2 templates besides the {{name}} wrappers. Can we use a string formatted here?

File: src/tests/test_command.py [hunk 8]
```diff
@@ -78,7 +78,8 @@ def test_parsed_sql_when_using_with(ip, sql_magic):
         sql_magic, ip.user_ns, line="--with author_one", cell="SELECT * FROM author_one"
     )
 
-    sql = "WITH author_one AS (SELECT * FROM author LIMIT 1) SELECT * FROM author_one"
+    sql = "WITH `author_one` AS (\n\n        SELECT * FROM author LIMIT 1\n        )\n\
```
Comment (yafimvo) [inline src/tests/test_command.py:81]: Why did you change the original query? Can we parametrize it and test both?

File: src/tests/test_connection.py [hunk 9]
```diff
@@ -36,15 +38,134 @@ def test_alias(cleanup):
...
+        monkeypatch (fixture): A convenient fixture for monkey-patching
+        cur_dialect (bool): Patched dialect name
+        expected_support_backtick (bool): Excepted boolean value to indicate
+        if the dialect supports backtick identifier
+    """
+    monkeypatch.setattr(Connection, "_get_curr_sqlglot_dialect", lambda: cur_dialect)
+    assert Connection.is_use_backtick_template() == expected_support_backtick
+
+
+def test_is_use_backtick_template_sqlglot_missing_dialect(monkeypatch):
```
Comment (yafimvo) [inline src/tests/test_connection.py:139]: Since `is_use_backtick_template` returns False for every exception, let's add a description/change the test name indicating which exception should be raised (so later, when we'll want to handle it differently - it will be easy to modify).

**Review by yafimvo — CHANGES_REQUESTED** [review]
Added minor notes.

General question - I see you're using this logic 

```
identifier = "`" if is_dialect_support_backtick else '"'
```

Is there any chance the default identifier (`"`) won't work?

**Comment by tonykploomber** [comment]
> Added minor notes.
> 
> General question - I see you're using this logic
> 
> ```
> identifier = "`" if is_dialect_support_backtick else '"'
> ```
> 
> Is there any chance the default identifier (`"`) won't work?

Good question, I don't have fully confidence to answer you 100% yes or no.

But originally before we use sqlglot, we use simple with_template includes `"`
```
with_template = Template(
    """WITH{% for name in with_ %} "{{name}}" AS ({{saved[name]._query}})\
{{ "," if not loop.last }}{% endfor %}{{query}}"""
)
```

We haven't met any issue then, now in this PR we just detect if the dialect support ` and we use backtick template, I think the impact should be not much

File: src/sql/connection.py [hunk 10]
```diff
@@ -365,14 +369,54 @@ def _get_curr_connection_info(cls):
...
+        except (ValueError, AttributeError, TypeError):
+            return False
+
+    @classmethod
+    def _transpile_query(cls, query):
+        """Translate the given SQL clause that's compatible to current connected
+        dialect by sqlglot
+
+        Args:
+            query (str): Original SQL clause
```
[the remark this answers is not in the document]
Comment (tonykploomber) [reply src/sql/connection.py:413]: Interesting, seems there are different formats for docstrings. I'll follow the style you suggest
https://numpydoc.readthedocs.io/en/latest/format.html

[the remark this answers is not in the document]
Comment (tonykploomber) [reply src/sql/connection.py:413]: The one I used is suggested by IDE, probably it's google style
https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings

[the remark this answers is not in the document]
Comment (tonykploomber) [reply src/sql/connection.py:413]: Ref: https://www.softwaretestinghelp.com/python-docstring-tutorial/

File: src/tests/test_command.py (hunk 8 above)
Comment (tonykploomber) [reply src/tests/test_command.py:81]: I think because it's we are using sqlite engine to test, and sqlite actually [support backtick symbol](https://github.com/tobymao/sqlglot/blob/main/sqlglot/dialects/sqlite.py#L27), so we need to change the test case here

**Comment by edublancas** [comment]
what's the status of this? This is blocking the 0.7 release so I'd like to merge this as soon as we can.

**Review by edublancas — APPROVED** [review]
